core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1488,7 +1488,7 @@ class ControllerChangeHandler(controller: KafkaController, eventManager: Control
 }
 
 case class ReassignedPartitionsContext(var newReplicas: Seq[Int] = Seq.empty,
-                                       var reassignIsrChangeHandler: PartitionReassignmentIsrChangeHandler = null)
+                                       var reassignIsrChangeHandler: Option[PartitionReassignmentIsrChangeHandler] = None)
```

Comment:
Perhaps a more defensive way is to initialize reassignIsrChangeHandler in the constructor instead of resetting it later.

Reply:
Yes, I would prefer that too, but it wasn't straightforward and I wasn't sure if there was a good reason to do it in 2 steps. I'll try harder to see if I can initialize it in the constructor.

Reply:
There are two scenarios where we need to create `ReassignedPartitionsContext`:

1. `PartitionReassignment.process`. This is easy.

2. `KafkaController.initializePartitionReassignment`. This one is harder. The current flow is that we initialise the controller state in `initializeControllerContext` and then we start the reassignment in `maybeTriggerPartitionReassignment` (where we set the ISR change handler). If we want to keep this arrangement, it doesn't seem like there's a clean way to set `reassignIsrChangeHandler` during construction. Thoughts @junrao?

Reply:
To construct a reassignIsrChangeHandler, we just need to know the partition, right? Could we just pass in partition to ReassignedPartitionsContext?

Reply:
My bad, I had somehow conflated the creation of `PartitionReassignmentIsrChangeHandler` with its registration. Pushed an update, let me know how that looks.